core/src/main/java/org/jdbi/v3/core/config/ConfigRegistry.java

```diff
@@ -81,12 +81,45 @@ public <C extends JdbiConfig<C>> C get(Class<C> configClass) {
     }
 
     private Function<ConfigRegistry, JdbiConfig<?>> configFactory(Class<? extends JdbiConfig<?>> configClass) {
-        return configFactories.computeIfAbsent(configClass, klass ->
-                registry -> {
-                    var config = JdbiClassUtils.findConstructorAndCreateInstance(klass, JDBI_CONFIG_TYPES, registry);
```

Comment:
This unfortunately does not fit well into the new helper paradigm, so I mostly reverted it to how it was, since we take care to use different argument lists with `invokeExact`.

Reply:
@hgschmie , yes, we can benchmark this. I am fine to have benchmarks before changing sensitive code, but please hold yourself to that standard too - #2425 was merged with no benchmarking and no review, and introduces a significant performance regression. Please consider #2432 which demonstrates that `MH.invokeExact` is by far the fastest implementation.

I object to your characterization of this as unreadable. What parts could be written better to be more readable? Or, maybe you need to spend some time familiarizing yourself with `MethodHandle`? It is not the prettiest code but it does what it needs to do, and is relatively straightforward.

MH is the new way of doing things and will not be going away. I think at this point it should be the default in any performance-sensitive code over "classic" reflection, unless there is a specific reason not to use it. Also, always prefer `invokeExact` when possible.

I like the idea of improving the `JdbiClassUtils` to return a `MethodHandle` that is stored for later re-use. However, it is a little tricky because the returned MH would have different method signatures depending on whether a `ConfigRegistry` is expected as an argument or not. It is entirely possible to fix this as well, but at some point the helper method is not generic anymore and ends up looking very much like just moving this code to `JdbiClassUtils`, which doesn't really solve any problem. I played with this for a bit but did not make good progress on deriving a helper method that is both fast and actually re-usable. So for now, this just brings us back to the baseline we had before.

> In its current shape, this seems to be more an exercise in proving that this can be done somehow using MethodHandles at any cost. I want to understand first what the value is before reverting the change. The loss of readability is IMHO a problem.

Right, because using `MH.invokeExact` carries significant performance gains. So, I spent significant effort converting the old code to be much faster. It is a little frustrating to leave for one week to find all that work thrown in the trash. I am on board with the goal to improve readability, but there very much is a reason that this code was the way it was.

> Why can't the findConstructorAndCreateInstance method not be rewritten to iterate over the list of parameters using method handles in the same way it does right now for finding constructors?

This is the tricky bit, because then the returned MH will accept an unknown number of parameters. We'd need to return some composite result that has both the MH and which signature it should be called with, otherwise you run into `WrongMethodTypeException`.

Reply:
I guess what we could do is take the MethodHandle we want to return and compose with `MethodHandles.dropArguments` to avoid the caller needing to know what kind of call we actually make. I will briefly look into this, although this is very much not what I was planning to spend this morning on.

Reply:
I played around with this a bit and I think it would work, however there is actually only one place that we need both 1) MethodHandle performance and 2) accept multiple type signatures, and that is here in `ConfigRegistry`. So, I am thinking that we should defer further extracting a general helper until we have a second use site that actually needs it.